Fix line breaks / formatting on `export --append`

Followup for #1036 

Context: https://github.com/papis/papis/pull/1036#issuecomment-3039877652, https://github.com/papis/papis/pull/1036#issuecomment-3040499663, https://github.com/papis/papis/pull/1036#pullrequestreview-2991052748

To summarize: on `export --append`, the resulting outfile is badly formatted in most cases, specifically:

- `bibtex`: no newline at end-of-file, so appended text doesn't have a proper linebreak (see https://github.com/papis/papis/pull/1036#issuecomment-3039877652)
- `csl`: no newline (as above)
- `json`:
    1. no newline (as above)
    2. instead of extending existing list, starts a new one (see https://github.com/papis/papis/pull/1036#issuecomment-3040499663)
- `yaml`: doesn't insert `---` in between documents on append
- `typst`: can't export multiple docs (?)
    - otherwise quite sane, although I suspect it would look better with a blank line in between on append

>> typst: can't export multiple docs (?)
>
> You mean our exporter? The format itself can definitely support multiple documents.

Yes, does this work for you?

`papis export -f typst --all`

I get this (`--log DEBUG`):

```
[DEBUG] utils: Indexing folders in '/home/tuure/files/docs/papis'.
[DEBUG] utils: Retrieved 56 valid folders.
[DEBUG] utils: Finished in 36.2 ms.
[DEBUG] database.cache: Computing 'papis_id' for each document.
[DEBUG] database.cache: Loaded 56 documents.
[DEBUG] database.cache: Querying database for '.'.
[ERROR] commands.export: Failed to load 'typst' exporter. Cannot export to this format.
  ┆ Traceback (most recent call last):
  ┆   File "/usr/lib/python3.13/site-packages/papis/commands/export.py", line 88, in run
  ┆     .plugin(document for document in documents))
  ┆      ~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ┆   File "/usr/lib/python3.13/site-packages/papis/hayagriva.py", line 229, in exporter
  ┆     create_reference(doc): to_hayagriva(doc) for doc in documents
  ┆                            ~~~~~~~~~~~~^^^^^
  ┆   File "/usr/lib/python3.13/site-packages/papis/hayagriva.py", line 169, in to_hayagriva
  ┆     parent_known_keys = HAYAGRIVA_TYPE_PARENT_KEYS[htype]
  ┆                         ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^
  ┆ KeyError: ''
```

Also happens with any 2 documents. But a single document works.

> To expand a bit on this: to properly support `json`, it seems like we need to let the exporter do the file writing / appending. I'm inclined to suggest reworking the exporter plugins to be a little class similar to `papis.importer.Importer` that can do whatever it needs to. What did you have in mind?

Yeah something along those lines, will have to look into it/experiment to find out more details :)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -92,6 +92,10 @@ user can abort, skip problematic moves, or fix collisions interactively. A new
 - Fix crash in `papis add` when a given file path contains spaces (the DOI
   importer no longer tries to resolve local paths on doi.org)
   ([#1201](https://github.com/papis/papis/issues/1201)).
+- The YAML exporter now starts each document with an explicit `---` marker, so
+  that `papis export --format yaml --append` produces a valid multi-document
+  file instead of merging the documents together
+  ([#1038](https://github.com/papis/papis/issues/1038)).
 
 # VERSION v0.15.0 (February 8th, 2026)
 

--- a/papis/exporters/yaml.py
+++ b/papis/exporters/yaml.py
@@ -12,8 +12,12 @@ def exporter(documents: list[Document]) -> str:
 
     from papis.document import to_dict
 
+    # NOTE: 'explicit_start' adds a '---' marker before every document, so that
+    # the result can be safely concatenated with an existing YAML file, e.g. by
+    # 'papis export --append'.
     string = yaml.dump_all(
         [to_dict(document) for document in documents],
-        allow_unicode=True)
+        allow_unicode=True,
+        explicit_start=True)
 
     return str(string)

--- a/tests/commands/test_explore.py
+++ b/tests/commands/test_explore.py
@@ -64,6 +64,7 @@ def test_explore_yaml_cli(tmp_library: TemporaryLibrary) -> None:
         exported_yaml = fd.read()
 
     assert re.match(
+        r"---\n"
         r"author: K. Popper\n"
         r"doi: 10.1021/ct5004252\n"
         r"papis_id: .*\n"

--- a/tests/commands/test_export.py
+++ b/tests/commands/test_export.py
@@ -200,3 +200,23 @@ def test_export_folder_all_cli(tmp_library: TemporaryLibrary) -> None:
     for d in dirs:
         assert os.path.exists(d)
         assert os.path.isdir(d)
+
+
+def test_export_yaml_append(tmp_library: TemporaryLibrary) -> None:
+    from papis.commands.export import cli
+
+    cli_runner = PapisRunner()
+    outfile = os.path.join(tmp_library.tmpdir, "test_append.yaml")
+
+    for query in ("krishnamurti", "popper"):
+        result = cli_runner.invoke(
+            cli,
+            ["--append", "--format", "yaml", "--out", outfile, query])
+        assert result.exit_code == 0
+
+    from papis.yaml import yaml_to_list
+    data = yaml_to_list(outfile, raise_exception=True)
+
+    assert len(data) == 2
+    assert "Krishnamurti" in data[0]["author"]
+    assert "Popper" in data[1]["author"]

